pkg/ssh/suftp.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2023 ghostloda
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ssh
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+const (
+	promptpwd      string = "sudo password"
+	promptsuccess  string = "sudo success!"
+	prompterror    string = "sudo failed!!"
+	promptlength   int    = 13
+	sshdConfigPath        = "/etc/ssh/sshd_config"
+)
+
+func NewSudoSftpClient(conn *ssh.Client, sudopwd string, opts ...sftp.ClientOption) (*sftp.Client, error) {
+	s, err := conn.NewSession()
+	if err != nil {
+		return nil, err
+	}
+
+	sftpServerPath, err := getSftpServerPath(conn, sudopwd)
```

Comment:
and perhaps here should use a cache, we don't want to get sftp-server path from remote ssh client each time when creating sftp session.

Reply:
> and perhaps here should use a cache, we don't want to get sftp-server path from remote ssh client each time when creating sftp session.

Is sftp-client caching better?

Reply:
of course yes, but most of the methods needs refactored.

Reply:
Is it enough to add a cache under this function?

```
func (c *Client) newClientAndSftpClient(host string) (*ssh.Client, *sftp.Client, error) {
	var (
		sshClient  *ssh.Client
		sftpClient *sftp.Client
		err        error
	)
	sshClient, err = c.connect(host)
	if err != nil {
		return nil, nil, err
	}
	// create sftp client
	if c.Option.sudo || c.Option.user != defaultUsername {
		sftpClient, err = NewSudoSftpClient(sshClient, c.password)
	} else {
		sftpClient, err = sftp.NewClient(sshClient)
	}

	return sshClient, sftpClient, err
}
```

Reply:
How about add a new pr to add a client cache? Like this approach https://github.com/sealerio/sealer/blob/main/utils/ssh/connect.go#L142